README.md
#  archinfo

Returns a list of running processes and the architecture they are running under.

Apple M1/Apple Silicon/arm64 macOS can run x86_64 programs via Rosetta and most M1 systems currently (~March 2021) very likely run a mix of x86_64 and arm64 processes.

Activity Monitor can show the architecture, but command line tools such as `ps` and `top` do not due to Apple hiding the details of the proper `sysctl()` incantations necessary to get this info.

Patrick Wardle reverse-engineered Activity Monitor — <https://www.patreon.com/posts/45121749> — and I slapped that hack into a bare-bones command line utility `archinfo`.

It returns columnar output or JSON (via `--json`) — that will work nicely with `jq` — of running processes and their respective architectures.

Build from source or grab from the releases.

```bash
$ archinfo
...
   5949  arm64 /System/Library/Frameworks/AudioToolbox.framework/AudioComponentRegistrar
   5923  arm64 /System/Library/CoreServices/LocationMenu.app/Contents/MacOS/LocationMenu
   5901 x86_64 /Library/Application Support/Adobe/Adobe Desktop Common/IPCBox/AdobeIPCBroker.app/Contents/MacOS/AdobeIPCBroker
   5873  arm64 /Applications/Utilities/Adobe Creative Cloud Experience/CCXProcess/CCXProcess.app/Contents/MacOS/../libs/Adobe_CCXProcess.node
   5863  arm64 /bin/sleep
   5861 x86_64 /Applications/Tailscale.app/Contents/PlugIns/IPNExtension.appex/Contents/MacOS/IPNExtension
   5855 x86_64 /Applications/Elgato Control Center.app/Contents/MacOS/Elgato Control Center
   5852 x86_64 /Applications/Tailscale.app/Contents/MacOS/Tailscale
   5849  arm64 /System/Library/CoreServices/TextInputSwitcher.app/Contents/MacOS/TextInputSwitcher
...
```

```bash
$ archinfo --pid $(pgrep keyboardservicesd)
  60298 x86_64 /usr/libexec/keyboardservicesd

$ archinfo --json --pid $(pgrep keyboardservicesd)
{"pid":60298,"arch":"x86_64","name":"/usr/libexec/keyboardservicesd"}
```

```
Rscript -e 'table(jsonlite::stream_in(textConnection(system("/usr/local/bin/archinfo --json", intern=TRUE)))$arch)'
##
## arm64 x86_64
##   419     29
```


NEWS.md
* 0.4.0 • 2021-05-25
- added options for only showing X86_64 or ARM64 processes
# 0.3.0 • 2021-03-14
- added option for retrieving process info for a single pid
- added more tests
- added leak check Makefile option
# 0.2.0 • 2021-03-14
- removed Xcode dependency
- added codesigning option
- added option for either columnar output or ndjson output

# 0.1.0 • 2021-03-13
- initial release

archinfo/procHelper.m
#import <sys/sysctl.h>
#import <Foundation/Foundation.h>

#include "procHelper.h"

// archInfo() is due to the spelunking work of Patrick Wardle <https://www.patreon.com/posts/45121749>
NSString* archInfo(pid_t pid) {
    
  size_t size;
  cpu_type_t type = -1;
  int mib[CTL_MAXNAME] = {0};
  size_t length = CTL_MAXNAME;
  struct kinfo_proc procInfo = {0};
  if (noErr != sysctlnametomib("sysctl.proc_cputype", mib, &length)) return(@"unknown");

  mib[length] = pid;
  length++;
  size = sizeof(cpu_type_t);
  if (noErr != sysctl(mib, (u_int)length, &type, &size, 0, 0)) return(@"unknown");
  
  if (CPU_TYPE_X86_64 == type) return(@"x86_64");
  if (CPU_TYPE_ARM64 == type) {
    mib[0] = CTL_KERN;  //(re)init mib
    mib[1] = KERN_PROC;
    mib[2] = KERN_PROC_PID;
    mib[3] = pid;
    
    length = 4; //(re)set length
    
    size = sizeof(procInfo); //(re)set size
    
    if(noErr != sysctl(mib, (u_int)length, &procInfo, &size, NULL, 0)) return(@"arm64"); //get proc info
    
    //'P_TRANSLATED' set? set architecture to 'Intel'
    return( (P_TRANSLATED == (P_TRANSLATED & procInfo.kp_proc.p_flag)) ? @"x86_64" : @"arm64");
    
  }
  
  return(@"unknown");
  
}

// The below is modified code from: <https://gist.github.com/s4y/1173880/9ea0ed9b8a55c23f10ecb67ce288e09f08d9d1e5>
//
// Copyright (c) 2020 DeepTech, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
// documentation files (the "Software"), to deal in the Software without restriction, including without limitation
// the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and
// to permit persons to whom the Software is furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO
// THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
// TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//

NSArray* allProcesses(){
  
  static int maxArgumentSize = 0;
  
  if (maxArgumentSize == 0) {
    size_t size = sizeof(maxArgumentSize);
    if (sysctl((int[]) { CTL_KERN, KERN_ARGMAX }, 2, &maxArgumentSize, &size, NULL, 0) == -1) {
      perror("sysctl argument size");
      maxArgumentSize = 4096; // Default
    }
  }
  
  NSMutableArray *processes = [NSMutableArray array];
  int mib[3] = { CTL_KERN, KERN_PROC, KERN_PROC_ALL };
  struct kinfo_proc *info;
  size_t length;
  int count;
  
  if (sysctl(mib, 3, NULL, &length, NULL, 0) < 0) return nil;
  if (!(info = malloc(length))) return nil;
  if (sysctl(mib, 3, info, &length, NULL, 0) < 0) {
    free(info);
    return nil;
  }
  
  count = (int)(length / sizeof(struct kinfo_proc));
  
  for (int i = 0; i < count; i++) {
    pid_t pid = info[i].kp_proc.p_pid;
    if (pid == 0) continue;
    size_t size = maxArgumentSize;
    char* buffer = (char *)malloc(length);
    if (sysctl((int[]){ CTL_KERN, KERN_PROCARGS2, pid }, 3, buffer, &size, NULL, 0) == 0) {
      
      NSString* executable = [NSString stringWithCString:(buffer+sizeof(int)) encoding:NSUTF8StringEncoding];
      [processes addObject:[NSDictionary dictionaryWithObjectsAndKeys:
                            [NSNumber numberWithInt:pid], @"pid",
                            executable, @"executable",
                            archInfo(pid), @"arch",
                            nil]];
    }
    
    free(buffer);

  }
  
  free(info);
  
  return processes;
}

NSString *processListToJSON(NSArray * processListDict) {
  NSError * err;
  NSData * jsonData = [NSJSONSerialization dataWithJSONObject:processListDict options:0 error:&err];
  return([[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding]);
}


archinfo.c
#include <unistd.h>
#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <mach/machine.h>
#include <sys/types.h>
#include <sys/sysctl.h>
#include <libgen.h>

#define VERSION "0.4.0"

#define noErr 0
#define DEFAULT_BUFFER_SIZE 4096
#define SYSCTL_ERROR 1

#define is_match(x) (x == 0)

enum fmt { JSON=0, COLUMNS };
enum show { ALL=0, ARM64, X86_64 };

static int max_arg_size = 0;

typedef struct ProcInfo {
  bool ok;
  char *name;
  char arch[10];  
} procinfo;

// arch_info() is due to the spelunking work of Patrick Wardle <https://www.patreon.com/posts/45121749>
static char *arch_info(pid_t pid) {

  size_t size;
  cpu_type_t type = -1;
  int mib[CTL_MAXNAME] = {0};
  size_t length = CTL_MAXNAME;
  struct kinfo_proc procInfo = {0};

  if (noErr != sysctlnametomib("sysctl.proc_cputype", mib, &length)) return("unknown");

  mib[length] = pid;
  length++;
  size = sizeof(cpu_type_t);

  if (noErr != sysctl(mib, (u_int)length, &type, &size, 0, 0)) return("unknown");
  
  if (CPU_TYPE_X86_64 == type) return("x86_64");

  if (CPU_TYPE_ARM64 == type) {

    mib[0] = CTL_KERN;  //(re)init mib
    mib[1] = KERN_PROC;
    mib[2] = KERN_PROC_PID;
    mib[3] = pid;
    
    length = 4; //(re)set length
    
    size = sizeof(procInfo); //(re)set size
    
    if(noErr != sysctl(mib, (u_int)length, &procInfo, &size, NULL, 0)) return("arm64"); //get proc info
    
    //'P_TRANSLATED' set? set architecture to 'x86_64'
    return( (P_TRANSLATED == (P_TRANSLATED & procInfo.kp_proc.p_flag)) ? "x86_64" : "arm64");
    
  }

  return("unknown");

}

// retrieve process info

procinfo proc_info(pid_t pid) {
 
  size_t size = max_arg_size;
  procinfo p;
  int mib[3] = { CTL_KERN, KERN_PROCARGS2, pid };
  struct kinfo_proc *info;
  size_t length;
  int count;

  // get size for buffer
  (void)sysctl(mib, 3, NULL, &length, NULL, 0);

  char* buffer = (char *)calloc(length+32, sizeof(char)); // need +32 b/c this is busted on Big Sur

  // get the info
  if (sysctl(mib, 3, buffer, &size, NULL, 0) == 0) {
    
    // copy the info
    p.ok = TRUE;
    p.name = buffer;
    strncpy(p.arch, arch_info(pid), 10);

  } else {
    free(buffer);
    p.ok = FALSE;
  }

  return(p);

}

// output one line of process info with architecture info

void output_one(enum fmt output_type, pid_t pid, procinfo p, bool only_basename) {
  if (output_type == COLUMNS) {
    printf(
      "%7d %6s %s\n", 
      pid,
      p.arch, 
      (only_basename ? basename((char *)(p.name+sizeof(int))) : p.name+sizeof(int))
    );
  } else if (output_type == JSON) {
    printf(
      "{\"pid\":%d,\"arch\":\"%s\",\"name\":\"%s\"}\n", 
      pid, 
      p.arch, 
      (only_basename ? basename((char *)(p.name+sizeof(int))) : p.name+sizeof(int))
    );
  }
}

// walk through process list, get PID and name, then pass that on to output_one() to
// grab the architecture and do the actual output

int enumerate_processes(enum fmt output_type, enum show to_show, bool only_basename) {

  int mib[3] = { CTL_KERN, KERN_PROC, KERN_PROC_ALL };
  struct kinfo_proc *info;
  size_t length;
  int count;

  // get the running process list
  if (sysctl(mib, 3, NULL, &length, NULL, 0) < 0) return(SYSCTL_ERROR);

  // make some room for results
  if (!(info = calloc(length, sizeof(char)))) return(SYSCTL_ERROR);

  // get the results
  if (sysctl(mib, 3, info, &length, NULL, 0) < 0) {
    free(info);
    return(SYSCTL_ERROR);
  }

  // how many results?
  count = (int)(length / sizeof(struct kinfo_proc));
  
  // for each result
  for (int i = 0; i < count; i++) {

    pid_t pid = info[i].kp_proc.p_pid; // get PID

    if (pid == 0) continue;

    procinfo p = proc_info(pid); // get process info

    if (p.ok) {
      if (
           (to_show == ALL) || 
           ((to_show == ARM64) && is_match(strncmp("arm", p.arch, 3))) || 
           ((to_show == X86_64) && is_match(strncmp("x86", p.arch, 3)))
         ) {
        output_one(output_type, pid, p, only_basename);
      }
      free(p.name);
    }
      
  }
  
  free(info);
  
  return(0);

}

// call to display cmdline tool help

void help() {

  printf("archinfo %s\n", VERSION);
  printf("boB Rudis <bob@rud.is>\n");
  printf("\n");
  printf("archinfo outputs a list of running processes with architecture (x86_64/amd64) information\n");
  printf("\n");
  printf("USAGE:\n");
  printf("    archinfo [--arm|--x86] [--basename] [--columns|--json] [--pid #]\n");
  printf("\n");
  printf("FLAGS/OPTIONS:\n");
  printf("    --arm             Only show arm64 processes (default is to show both)\n");
  printf("    --x86             Only show x86_64 processes  (default is to show both)\n");
  printf("    --basename        Only show basename of process executable\n");
  printf("    --columns         Output process list in columns (default)\n");
  printf("    --json            Output process list in ndjson\n");
  printf("    --pid #           Output process architecture info for the specified process\n");
  printf("    --help            Display this help text\n");

}

void init() {
   
  if (max_arg_size == 0) {
    size_t size = sizeof(max_arg_size);
    if (sysctl((int[]) { CTL_KERN, KERN_ARGMAX }, 2, &max_arg_size, &size, NULL, 0) == -1) {
      perror("sysctl argument size");
      max_arg_size = DEFAULT_BUFFER_SIZE;
    }
  }

}

int main(int argc, char** argv) {

  int c;
  enum show to_show = ALL;
  bool show_help = FALSE;
  bool do_one = FALSE;
  bool only_basename = FALSE;
  pid_t pid = -1;
  enum fmt output_type = COLUMNS;

  while(true) {

    int this_option_optind = optind ? optind : 1;
    int option_index = 0;

    static struct option long_options[] = {
      { "arm",      no_argument,       0,  'a' },
      { "x86",      no_argument,       0,  'x' },
      { "basename", no_argument,       0,  'b' },
      { "json",     no_argument,       0,  'j' },
      { "columns",  no_argument,       0,  'c' },
      { "help",     no_argument,       0,  'h' },
      { "pid",      required_argument, 0,  'p' },
      { 0,          0,                 0,  0 }
    };

    c = getopt_long(argc, argv, "axbjchp:", long_options, &option_index);

    if (c == -1) break;

    switch(c) {

      case 'a': to_show = ARM64; break;
      case 'x': to_show = X86_64; break;
      case 'b': only_basename = TRUE; break;
      case 'p': do_one = TRUE; pid = atoi(optarg); break;
      case 'h': show_help = TRUE; break;
      case 'j': output_type = JSON; break;
      case 'c': output_type = COLUMNS; break;

    }

  }

  init();

  // only show help if --help is in the arg list

  if (show_help) {
    help();
    return(0);
  }

  // only do one process if --pid is in the arg list

  if (do_one) {
    if (pid > 0) {
      procinfo p = proc_info(pid);
      if (p.ok) {
        output_one(output_type, pid, p, only_basename);
        free(p.name);
        return(0);
      }
    }
    return(SYSCTL_ERROR);
  } 

 // otherwise do them all

  return(enumerate_processes(output_type, to_show, only_basename));

}
